File: libs/vgc/core/tests/test_arrays.py
#!/usr/bin/python3

# Copyright 2021 The VGC Developers
# See the COPYRIGHT file at the top-level directory of this distribution
# and at https://github.com/vgc/vgc/blob/master/COPYRIGHT
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

from vgc.core import (
    IndexError,
    DoubleArray,
    FloatArray,
    IntArray,
    SharedConstIntArray
)

def get1DArrayClasses():
    return [DoubleArray, FloatArray, IntArray]

# Note: we use exactly representable floating-point values (e.g: 1.5 = 1 + 1/2)
# to avoid comparison errors when converting from Python's float (64bit) to
# vgc.core.FloatArray (32bit), back to Python's float for the comparison.

def get1DValues():
    return [0.5, 0.5, 42]

def get1DValuess():
    return [[1.5, 0.5, 2.5], [1.5, 0.5, 2.5], [42, 43, 41]]

def testDefaultConstructor_(self, Array):
    a = Array()
    self.assertEqual(len(a), 0)

def testZeroInitializingConstructor_(self, Array):
    n = 3
    a = Array(n)
    self.assertEqual(len(a), n)
    for x in a:
        self.assertEqual(x, 0)

def testValueInitializingConstructor_(self, Array, value):
    n = 3
    a = Array(n, value)
    self.assertEqual(len(a), n)
    for x in a:
        self.assertEqual(x, value)

def assertArrayEqualToList_(self, a, l):
    self.assertEqual(len(a), len(l))
    for i in range(len(a)):
        self.assertEqual(a[i], l[i])

def testListInitializingConstructor_(self, Array, values):
    a = Array(values)
    assertArrayEqualToList_(self, a, values)

def testPrepend_(self, Array, values):
    a = Array()
    for v in values[::-1]:
        a.prepend(v)
    assertArrayEqualToList_(self, a, values)

def testAppend_(self, Array, values):
    a = Array()
    for v in values:
        a.append(v)
    assertArrayEqualToList_(self, a, values)

def testPop_(self, Array, values):
    a = Array(values)
    b = list(values)
    x = a.pop()
    y = b.pop()
    self.assertEqual(x, y)
    assertArrayEqualToList_(self, a, b)

    a = Array(values)
    b = list(values)
    x = a.pop(0)
    y = b.pop(0)
    self.assertEqual(x, y)
    assertArrayEqualToList_(self, a, b)

    a = Array(values)
    b = list(values)
    x = a.pop(1)
    y = b.pop(1)
    self.assertEqual(x, y)
    assertArrayEqualToList_(self, a, b)

    a = Array(values)
    b = list(values)
    x = a.pop(-1)
    y = b.pop(-1)
    self.assertEqual(x, y)
    assertArrayEqualToList_(self, a, b)

    a = Array(values)
    b = list(values)
    x = a.pop(len(a)-1)
    y = b.pop(len(b)-1)
    self.assertEqual(x, y)
    assertArrayEqualToList_(self, a, b)

    a = Array(values)
    b = list(values)
    x = a.pop(-len(a))
    y = b.pop(-len(b))
    self.assertEqual(x, y)
    assertArrayEqualToList_(self, a, b)

    a = Array(values)
    b = list(values)
    x = a.pop(-1)
    y = b.pop(-1)
    self.assertEqual(x, y)
    assertArrayEqualToList_(self, a, b)

    a = Array(values)
    with self.assertRaises(IndexError):
        a.pop(-len(a)-1)
    with self.assertRaises(IndexError):
        a.pop(len(a))
    a = Array()
    with self.assertRaises(IndexError):
        a.pop()

class Test1DArrays(unittest.TestCase):

    def testDefaultConstructor(self):
        Arrays = get1DArrayClasses()
        for Array in Arrays:
           testDefaultConstructor_(self, Array)

    def testZeroInitializingConstructors(self):
        Arrays = get1DArrayClasses()
        for Array in Arrays:
           testZeroInitializingConstructor_(self, Array)

    def testValueInitializingConstructors(self):
        Arrays = get1DArrayClasses()
        values = get1DValues()
        for Array, value in zip(Arrays, values):
           testValueInitializingConstructor_(self, Array, value)

    def testListInitializingConstructors(self):
        Arrays = get1DArrayClasses()
        valuess = get1DValuess()
        for Array, values in zip(Arrays, valuess):
           testListInitializingConstructor_(self, Array, values)

    def testIndex(self):
        a = DoubleArray("[3.5, 4, 42.5, 1]")
        self.assertEqual(a.index(42.5), 2)

    def testPrepend(self):
        Arrays = get1DArrayClasses()
        valuess = get1DValuess()
        for Array, values in zip(Arrays, valuess):
           testPrepend_(self, Array, values)

    def testAppend(self):
        Arrays = get1DArrayClasses()
        valuess = get1DValuess()
        for Array, values in zip(Arrays, valuess):
           testAppend_(self, Array, values)

    def testInsert(self):
        a = DoubleArray("[3.5, 4, 42.5, 1]")
        a.insert(2, 42)
        a.insert(-4, 1)
        b = DoubleArray("[3.5, 1, 4, 42, 42.5, 1]")
        self.assertEqual(a, b)

    def testPop(self):
        Arrays = get1DArrayClasses()
        valuess = get1DValuess()
        for Array, values in zip(Arrays, valuess):
           testPop_(self, Array, values)

    def testParse(self):
        a = DoubleArray("[3.5, 42]")
        self.assertEqual(a, DoubleArray([3.5, 42]))
        a = FloatArray("[3.5, 42]")
        self.assertEqual(a, FloatArray([3.5, 42]))
        a = IntArray("[43, 42]")
        self.assertEqual(a, IntArray([43, 42]))
        a = DoubleArray("  [  3.5 \n,  42 ]   ")
        self.assertEqual(a, DoubleArray([3.5, 42]))
        a = DoubleArray("[]")
        self.assertEqual(a, DoubleArray())
        a = DoubleArray("  [ \n]   ")
        self.assertEqual(a, DoubleArray())

    def testFormat(self):
        a = DoubleArray("[3.5, 42, 1]")
        self.assertEqual(str(a), "[3.5, 42, 1]")

    def testSharedConst(self):
        a = IntArray([3, 1, 2])
        b = SharedConstIntArray(a)
        c = SharedConstIntArray(b)
        self.assertEqual(a, b)
        self.assertEqual(b, c)
        self.assertEqual(a, c)

if __name__ == '__main__':
    unittest.main()

File: libs/vgc/geometry/tests/test_arrays.cpp
// Copyright 2022 The VGC Developers
// See the COPYRIGHT file at the top-level directory of this distribution
// and at https://github.com/vgc/vgc/blob/master/COPYRIGHT
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <gtest/gtest.h>
#include <vgc/core/stopwatch.h>
#include <vgc/geometry/vec2d.h>

using vgc::Int;
using vgc::core::Stopwatch;
using vgc::geometry::Vec2d;
using vgc::geometry::Vec2dArray;

TEST(TestArrays, Resize) {
    Vec2dArray a = {{1, 2}, {3, 4}};
    Vec2dArray b = {{1, 2}};
    Vec2dArray c = {{1, 2}, {0, 0}};
    a.removeLast();
    EXPECT_EQ(a, b);
    a.resize(2);
    EXPECT_EQ(a, c);
}

TEST(TestArrays, ResizeNoInit) {
    Vec2dArray a = {{1, 2}, {3, 4}};
    Vec2dArray b = {{1, 2}};
    Vec2dArray c = {{1, 2}, {3, 4}};
    a.removeLast();
    EXPECT_EQ(a, b);
    a.resizeNoInit(2);
    EXPECT_EQ(a, c);
}

#ifndef VGC_DEBUG_BUILD

namespace {

void measureResizePerf(
    Vec2dArray& array,
    void (Vec2dArray::*resizeFn)(Int),
    Int n,
    double& elapsed,
    bool isWarmup) {

    vgc::core::Stopwatch s;
    (array.*resizeFn)(0);
    EXPECT_EQ(array.length(), 0); // Ensures that resize(0) isn't optimized out
    (array.*resizeFn)(n);
    EXPECT_EQ(array[n - 1], Vec2d()); // Ensures that resize(n) isn't optimized out
    if (!isWarmup) {
        elapsed += s.elapsed();
    }
}

} // namespace

TEST(TestArrays, ResizeNoInitPerf) {
    Int k = 1000;
    Int n = 100000;
    Vec2dArray a(n);
    double elapsedInit = 0;
    double elapsedNoInit = 0;
    Int numWarmupIterations = k / 10;
    for (Int i = 0; i < k; ++i) {
        bool isWarmup = (k < numWarmupIterations);
        measureResizePerf(a, &Vec2dArray::resize, n, elapsedInit, isWarmup);
        measureResizePerf(a, &Vec2dArray::resizeNoInit, n, elapsedNoInit, isWarmup);
    }
    vgc::core::print("Resize with init    = {:.7f} sec.\n", elapsedInit);
    vgc::core::print("Resize without init = {:.7f} sec.\n", elapsedNoInit);
    EXPECT_LT(elapsedNoInit, 0.01 * elapsedInit);

    // Performance results on DELL Precision 3561, i7-11850H, 32GB RAM:
    //    Resize with init    = 0.0343690 sec.
    //    Resize without init = 0.0000265 sec.
}

#endif // VGC_DEBUG_BUILD

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}
